File: applicatie/src/main/java/com/vdzon/administratie/factuur/FactuurService.java
package com.vdzon.administratie.factuur;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.vdzon.administratie.auth.SessionHelper;
import com.vdzon.administratie.crud.UserCrud;
import com.vdzon.administratie.dto.FactuurDto;
import com.vdzon.administratie.model.Bestelling;
import com.vdzon.administratie.model.Factuur;
import com.vdzon.administratie.model.Gebruiker;
import com.vdzon.administratie.pdfgenerator.GenerateFactuur;
import com.vdzon.administratie.util.SingleAnswer;
import spark.Request;
import spark.Response;

import javax.inject.Inject;
import java.io.BufferedOutputStream;

public class FactuurService {

    @Inject
    UserCrud crudService;

    protected Object putFactuur(Request req, Response res) throws Exception {
        try {
            String uuid = SessionHelper.getAuthenticatedUserUuid(req);
            Gebruiker gebruiker = crudService.getGebruiker(uuid);
            if (gebruiker == null) {
                res.status(404);
                return new SingleAnswer("not found");
            }
            String factuurJson = req.body();
            Factuur factuur = null;
            ObjectMapper mapper = new ObjectMapper();
            FactuurDto factuurDto = mapper.readValue(factuurJson, FactuurDto.class);
            factuur = factuurDto.toFactuur();

            removeFactuur(gebruiker, factuur.getUuid());
            addFactuur(gebruiker, factuur);


//            gebruiker.getDefaultAdministratie().removeFactuur(factuur.getUuid());
//            gebruiker.getDefaultAdministratie().addFactuur(factuur);
            crudService.updateGebruiker(gebruiker);
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        }
        return new SingleAnswer("ok");
    }

    protected Object removeFactuur(Request req, Response res) throws Exception {
        try {
            String uuid = SessionHelper.getAuthenticatedUserUuid(req);
            Gebruiker gebruiker = crudService.getGebruiker(uuid);
            if (gebruiker == null) {
                res.status(404);
                return new SingleAnswer("not found");
            }
            String factuurUuid = req.params(":uuid");
            if ("undefined".equals(factuurUuid)) {
                factuurUuid = null;
            }
            gebruiker.getDefaultAdministratie().removeFactuur(factuurUuid);
            crudService.updateGebruiker(gebruiker);
            return new SingleAnswer("ok");
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        }

    }


    private void addFactuur(Gebruiker gebruiker, Factuur factuur) {
        String nieuwBestellingNummer = factuur.getGekoppeldeBestellingNummer();
        Bestelling gekoppelBestelling = gebruiker.getDefaultAdministratie().getBestellingByBestellingNummer(nieuwBestellingNummer);
        if (gekoppelBestelling!=null){
            Bestelling updatedBestelling = new Bestelling(
                    gekoppelBestelling.getBestellingNummer(),
                    factuur.getFactuurNummer(),
                    gekoppelBestelling.getBestellingDate(),
                    gekoppelBestelling.getContact(),
                    gekoppelBestelling.getBestellingRegels(),
                    gekoppelBestelling.getUuid()
            );
            gebruiker.getDefaultAdministratie().removeBestelling(gekoppelBestelling.getUuid());
            gebruiker.getDefaultAdministratie().addBestelling(updatedBestelling);
        }
        gebruiker.getDefaultAdministratie().addFactuur(factuur);
    }

    private void removeFactuur(Gebruiker gebruiker, String uuid) {
        Factuur factuur = gebruiker.getDefaultAdministratie().getFactuur(uuid);
        String bestellingNummerOudeFactuur = factuur == null ? null : factuur.getGekoppeldeBestellingNummer();
        Bestelling gekoppelBestelling = gebruiker.getDefaultAdministratie().getBestellingByBestellingNummer(bestellingNummerOudeFactuur);
        if (gekoppelBestelling!=null){
            Bestelling updatedBestelling = new Bestelling(
                    gekoppelBestelling.getBestellingNummer(),
                    null,
                    gekoppelBestelling.getBestellingDate(),
                    gekoppelBestelling.getContact(),
                    gekoppelBestelling.getBestellingRegels(),
                    gekoppelBestelling.getUuid()
            );
            gebruiker.getDefaultAdministratie().removeBestelling(gekoppelBestelling.getUuid());
            gebruiker.getDefaultAdministratie().addBestelling(updatedBestelling);
        }
        gebruiker.getDefaultAdministratie().removeFactuur(uuid);
    }



    protected Object getPdf(Request req, Response res) throws Exception {
        try {
            String uuid = SessionHelper.getAuthenticatedUserUuid(req);
            Gebruiker gebruiker = crudService.getGebruiker(uuid);
            if (gebruiker == null) {
                res.status(404);
                return new SingleAnswer("not found");
            }
            String factuurUuid = req.params(":uuid");
            if ("undefined".equals(factuurUuid)) {
                res.status(404);
                return new SingleAnswer("not found");
            }

            Factuur factuur = gebruiker.getDefaultAdministratie().getFactuur(factuurUuid);
            res.raw().setContentType("application/pdf");
            res.raw().setHeader("Content-Disposition","attachment; filename="+factuur.getFactuurNummer()+".pdf");
            try (BufferedOutputStream zipOutputStream = new BufferedOutputStream(res.raw().getOutputStream())) {
                GenerateFactuur.buildPdf(gebruiker.getDefaultAdministratie(), factuur, zipOutputStream);
                zipOutputStream.flush();
                zipOutputStream.close();
            }

            return null;
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        }

    }

}

File: rest/impl/src/main/java/com/vdzon/administratie/rest/factuur/FactuurService.kt
package com.vdzon.administratie.rest.factuur

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.vdzon.administratie.database.UserDao
import com.vdzon.administratie.dto.FactuurDto
import com.vdzon.administratie.model.BoekingenCache
import com.vdzon.administratie.model.Factuur
import com.vdzon.administratie.model.Gebruiker
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetFactuur
import com.vdzon.administratie.pdfgenerator.factuur.GenerateFactuur
import com.vdzon.administratie.util.SessionHelper
import com.vdzon.administratie.util.SingleAnswer
import spark.Request
import spark.Response
import java.io.BufferedOutputStream
import javax.inject.Inject

class FactuurService
@Inject
constructor(val generateFactuur: GenerateFactuur, var daoService: UserDao) {

    @Throws(Exception::class)
    fun putFactuur(req: Request, res: Response): Any {
        val gebruiker = SessionHelper.getGebruikerOrThowForbiddenException(req, daoService)
        val factuurJson = req.body()
        var factuur: Factuur? = null
        val mapper = jacksonObjectMapper()
        val factuurDto = mapper.readValue(factuurJson, FactuurDto::class.java)
        factuur = factuurDto.toFactuur()

        removeFactuur(gebruiker, factuur.uuid)
        addFactuur(gebruiker, factuur)

        val boekingenVanFactuur = BoekingenCache(gebruiker.defaultAdministratie.boekingen).getBoekingenVanFactuur(factuur.factuurNummer)
        for (boeking in boekingenVanFactuur) {
            var found = false
            if (factuurDto.boekingen != null) {
                for (boekingDto in factuurDto.boekingen!!) {
                    if (boekingDto.uuid == boeking.uuid) {
                        found = true
                    }
                }
            }
            if (!found) {
                gebruiker.defaultAdministratie.removeBoeking(boeking.uuid)
            }
        }

        daoService.updateGebruiker(gebruiker)
        return SingleAnswer("ok")
    }

    @Throws(Exception::class)
    fun removeFactuur(req: Request, res: Response): Any {
        val gebruiker = SessionHelper.getGebruikerOrThowForbiddenException(req, daoService)
        var factuurUuid: String? = req.params(":uuid")
        if ("undefined" == factuurUuid) {
            factuurUuid = null
        }
        val factuur = gebruiker.defaultAdministratie.getFactuur(factuurUuid!!)

        gebruiker.defaultAdministratie.removeFactuur(factuurUuid)
        removeBoekingenVanFactuur(gebruiker, factuur!!.factuurNummer)

        daoService.updateGebruiker(gebruiker)
        return SingleAnswer("ok")
    }


    private fun addFactuur(gebruiker: Gebruiker, factuur: Factuur) {
        val nieuwBestellingNummer = factuur.gekoppeldeBestellingNummer
        val gekoppelBestelling = gebruiker.defaultAdministratie.getBestellingByBestellingNummer(nieuwBestellingNummer)
        if (gekoppelBestelling != null) {
            val updatedBestelling = gekoppelBestelling.copy(gekoppeldFactuurNummer = factuur.factuurNummer)
            gebruiker.defaultAdministratie.removeBestelling(gekoppelBestelling.uuid)
            gebruiker.defaultAdministratie.addBestelling(updatedBestelling)
        }
        gebruiker.defaultAdministratie.addFactuur(factuur)
    }

    private fun removeFactuur(gebruiker: Gebruiker, uuid: String) {
        val factuur = gebruiker.defaultAdministratie.getFactuur(uuid)
        val bestellingNummerOudeFactuur = factuur?.gekoppeldeBestellingNummer
        val gekoppelBestelling = gebruiker.defaultAdministratie.getBestellingByBestellingNummer(bestellingNummerOudeFactuur)
        if (gekoppelBestelling != null) {
            val updatedBestelling = gekoppelBestelling.copy(gekoppeldFactuurNummer = null)
            gebruiker.defaultAdministratie.removeBestelling(gekoppelBestelling.uuid)
            gebruiker.defaultAdministratie.addBestelling(updatedBestelling)
        }
        gebruiker.defaultAdministratie.removeFactuur(uuid)
    }

    private fun removeBoekingenVanFactuur(gebruiker: Gebruiker, factuurNr: String) {
        val defaultAdministratie = gebruiker.defaultAdministratie
        defaultAdministratie.boekingen
                .filter { boeking -> boeking is BoekingMetFactuur }
                .map { boeking -> boeking as BoekingMetFactuur }
                .filter { boeking -> boeking.factuurNummer == factuurNr }
                .forEach { boeking -> defaultAdministratie.removeBoeking(boeking.uuid) }
    }

    @Throws(Exception::class)
    fun getPdf(req: Request, res: Response): Any? {
        val gebruiker = SessionHelper.getGebruikerOrThowForbiddenException(req, daoService)
        val factuurUuid = req.params(":uuid")
        if ("undefined" == factuurUuid) {
            res.status(404)
            return SingleAnswer("not found")
        }

        val factuur = gebruiker.defaultAdministratie.getFactuur(factuurUuid)
        res.raw().contentType = "application/pdf"
        res.raw().setHeader("Content-Disposition", "attachment; filename=" + factuur!!.factuurNummer + ".pdf")
        BufferedOutputStream(res.raw().outputStream).use { zipOutputStream ->
            generateFactuur.buildPdf(gebruiker.defaultAdministratie, factuur, zipOutputStream)
            zipOutputStream.flush()
            zipOutputStream.close()
        }
        return null
    }

}
